### PR TITLE
fix logger creating directory structure too early in DDP

### LINKED ISSUE #6364: DDP: Multiple processes try to create the logger directory tree
## 🐛 Bug

An user from our supercomputing center run into an issue which I think turned out to be a bug in PyTorch-Lightning.

When using the DDP accelerator together with a logger, multiple processes will try creating the logger directory tree, causing some errors about already existing directories or files.

## Troubleshooting

PyTorch-Lightning uses extensively the `rank_zero_only` function to ensure that some actions are only performed by the process with rank 0:
https://github.com/PyTorchLightning/pytorch-lightning/blob/b3b8f95e2a1ac040f6ff8f848542a1e5a27edfee/pytorch_lightning/utilities/distributed.py#L35-L42

`rank_zero_only.rank` default value is set there:
https://github.com/PyTorchLightning/pytorch-lightning/blob/b3b8f95e2a1ac040f6ff8f848542a1e5a27edfee/pytorch_lightning/utilities/distributed.py#L45-L46
but can be set in other modules, for example in our case DDP:
https://github.com/PyTorchLightning/pytorch-lightning/blob/b3b8f95e2a1ac040f6ff8f848542a1e5a27edfee/pytorch_lightning/plugins/training_type/ddp.py#L227-L228

Unfortunately it seems that the initialization by the DDP module happens too late, I think because of commit da6dbc8d1d128cf783d7151b012a5502bbd52bf5:
- `self.setup_trainer(model)` gets called on line 467 effectively initializing the logger and creating the logger directory tree
- DDP initialization and thus `rank_zero_only.rank` getting the correct value only happens at line 477 when calling `self.training_type_plugin.pre_training()`.

### To Reproduce

I have attached the code the user provided together the Slurm script: [only_rank_zero.tar.gz](https://github.com/PyTorchLightning/pytorch-lightning/files/6092665/only_rank_zero.tar.gz).

I understand that you would prefer a `BoringModel` and Collab based reproducer but I am from the HPC world and I am not used to those. Let me know if I can help in any other way. I hope that my own digging into the code will hep.

### Environment (probably not relevant in this case)

 - PyTorch Version: 1.7.1
 - OS: Linux (Red Hat 8.1)
 - How you installed PyTorch: `conda`, I tried the latest version of  PyTorch-Lightning available on `conda` but also tested installing the current master branch from source and the behavior is still the same.
 - Python version: 3.7.10
 - CUDA/cuDNN version: 11.0.221/8.0.5
 - GPU models and configuration: NVIDIA V100

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #6364

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/trainer/logging_/test_distributed_logging.py [hunk 1]
```diff
@@ -66,3 +67,37 @@ def test_global_zero_only_logging_ddp_spawn(tmpdir):
...
+class LoggerCallsObserver(Callback):
+
+    def on_before_accelerator_backend_setup(self, trainer, pl_module):
+        assert not trainer.logger.method_calls
+        assert not os.listdir(trainer.logger.save_dir)
+
+    def on_train_start(self, trainer, pl_module):
+        assert trainer.logger.method_call
+        trainer.logger.log_hyperparams.assert_called_once()
+        trainer.logger.log_graph.assert_called_once()
```
Comment (carmocca) [inline tests/trainer/logging_/test_distributed_logging.py:81]: nit: can you put this class inside the test? makes it easier to track that is only used in one place

**Review by carmocca — APPROVED** [review]
Great! Does the test fail in current master?

File: tests/trainer/logging_/test_distributed_logging.py (hunk 1 above)
Comment (awaelchli) [reply tests/trainer/logging_/test_distributed_logging.py]: moved :)

**Comment by awaelchli** [comment]
@carmocca yes, it fails on master :)

**Review by ananthsub — APPROVED** [review]

**Review by s-rog — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -509,6 +493,13 @@ def fit(
     def pre_dispatch(self):
         self.accelerator.pre_dispatch()
 
+        # log hyper-parameters
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:496]: Can we wrap this into a function ?

**Review by tchaton — APPROVED** [review]
LGTM !
